[RFC,WIP] implement PQC PSK key exchange handshake

jonasjelonek opened the pull request:
implements PQC PSK key exchange for unetd

**Just as RFC for the handshake/key exchange process and how it's implemented.**

This is still missing proper integration into unetd's internals, especially pex message send/recv flow.
This doesn't compile nor run yet.

nbd168 committed:
lib: add utility function for getting random data

Try getrandom/arc4random first, only fall back to /dev/urandom when necessary

Signed-off-by: Felix Fietkau <nbd@nbd.name>

nbd168 committed:
lib: add sntrup761 code

This will be used for quantum safe handshakes

Signed-off-by: Felix Fietkau <nbd@nbd.name>

nbd168 committed:
chacha20.h: add missing includes

Signed-off-by: Felix Fietkau <nbd@nbd.name>

nbd168 committed:
lib: add shake / fips202

Used for ML-DSA

Signed-off-by: Felix Fietkau <nbd@nbd.name>

nbd168 committed:
lib: add ML-DSA 44 code

Used for post-quantum signatures

Signed-off-by: Felix Fietkau <nbd@nbd.name>

nbd168 committed:
mldsa: implement function for deriving public key from secret key

Signed-off-by: Felix Fietkau <nbd@nbd.name>

nbd168 committed:
sntrup761: add function for deriving public key from secret key

Signed-off-by: Felix Fietkau <nbd@nbd.name>

nbd168 committed:
mldsa: add support for passing seed to the keygen

Signed-off-by: Felix Fietkau <nbd@nbd.name>

nbd168 committed:
cli: add support for generating pq host/signing keys

Signed-off-by: Felix Fietkau <nbd@nbd.name>

nbd168 commented on wg-linux.c:
```diff
@@ -233,6 +233,7 @@ wg_linux_peer_update(struct network *net, struct network_peer *peer, enum wg_upd
 	}
 
 	nla_put_u32(req.msg, WGPEER_A_FLAGS, WGPEER_F_REPLACE_ALLOWEDIPS);
+	nla_put(req.msg, WGPEER_A_PRESHARED_KEY, WG_KEY_LEN, peer->psk);
```
Please add PSK only if PQ crypto is being used between peers

jonasjelonek replied:
will fix it

nbd168 commented on psk-kex.h:
```diff
@@ -0,0 +1,61 @@
+	uint8_t c2_mac[MAC_LEN];
+	uint8_t c3_mac[MAC_LEN];
+	uint8_t nonce[CHACHA20_NONCE_SIZE];
+};
```
Please split up messages to stay below payload size of 1280 bytes.

jonasjelonek replied:
will fix it

nbd168 commented on host.c:
```diff
@@ -160,6 +170,15 @@ network_host_create(struct network *net, struct blob_attr *attr, bool dynamic)
 
+	pqc_key_buf = blobmsg_get_string(tb[NETWORK_HOST_PQC_KEY]);
+	if (pqc_key_buf) {
+        strsep(&pqc_key_buf, ":");
```
The network file should not have multiple keys in the pqc_key option. The keygen output should have been split up before being put into the JSON.

jonasjelonek replied:
This - in combination with your following comment - implements @aparcar's idea to have some kind of identifier/type in front of the key like `sntrup761-XXX:<key>`. But we can skip this for now.

nbd168 commented on host.c:
```diff
@@ -119,11 +127,13 @@ network_host_create(struct network *net, struct blob_attr *attr, bool dynamic)
 	struct blob_attr *tb[__NETWORK_HOST_MAX];
 	struct blob_attr *cur, *ipaddr, *subnet, *meta;
 	uint8_t key[CURVE25519_KEY_SIZE];
+	uint8_t pqc_key[32 + SNTRUP761_PUB_SIZE] = {0};
```
Why the extra 32 bytes?

jonasjelonek replied:
See comment above, 32 bytes was just a wild guess how long such an identifier/type could be.

nbd168 commented on CMakeLists.txt:
```diff
@@ -36,18 +36,18 @@ ELSE()
-  SET(udebug "")
+	SET(ubus "")
+	SET(udebug "")
 ENDIF()
```
Those changes should be dropped or put into separate commits.

jonasjelonek replied:
will fix it

nbd168 reviewed (commented):
Seems like a good start! Added a few comments...

aparcar commented on CMakeLists.txt:
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.3)
+cmake_minimum_required(VERSION 3.5)
```
Please put this in it's own commit, would be nice to not carry this in my downstream setup ;)

jonasjelonek committed:
WIP: PQC PSK key exchange handshake

implements PQC PSK key exchange for unetd

THIS IS UNFINISHED AND DOES NOT WORK YET!

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>

jonasjelonek commented on psk-kex.c:
```diff
@@ -0,0 +1,368 @@
+	return last == 0 || now - last > HANDSHAKE_INTERVAL;
+}
+
+/* This should be called regularly by some kind of timer */
```
@nbd168 should this get its own timer or can we reuse one of the existing timers like `request_update_timer` or `connect_timer`?

nbd168 replied:
I think it should get its own timer.

jonasjelonek commented on psk-kex.c:
```diff
@@ -0,0 +1,368 @@
+	resp->last_handshake_time = peer->state.last_psk_handshake;
+	resp->need_handshake = psk_kex_need_handshake(peer);
+
+	pex_msg_send_ext(net, peer, /* address outside of tunnel */ addr);
```
@nbd168 how may I get the address to send this to the "peer" outside of the tunnel on global pex port?

nbd168 replied:
You should go through all addresses in peer->state.next_endpoint, possibly checking for duplicates.

jonasjelonek replied:
For my understanding:
- what's the point of the different endpoint types?
- why couldn't I just use `peer->state.endpoint` or `peer->state.next_endpoint[ENDPOINT_TYPE_PEX]` as destination?

nbd168 replied:
`ENDPOINT_TYPE_PEX` is an endpoint address received from another host.
`ENDPOINT_TYPE_ENDPOINT_NOTIFY` and `ENDPOINT_TYPE_ENDPOINT_PORT_NOTIFY` are endpoint addresses received from peers directly (former from the PEX port, latter from the wireguard port).

If you don't have any peer connection yet, the `ENDPOINT_TYPE_PEX` will not be filled. Peers for other hosts are only exchanged over the tunnel.
As for the other types, availability depends on which packets make it through, especially in the NAT case. When the basic work on the handshake is done, we may need to add some improvements to NAT handling, since there will be some limitations compared to non-PEX handshake.

jonasjelonek commented on psk-kex.c:
```diff
@@ -0,0 +1,368 @@
+	default: return;
+	}
+
+	// TBD: do we need to explicitly trigger a network update to use the new PSK or does that work automatically?
```
@nbd168 do you know?

nbd168 replied:
You can do a wg_peer_update(net, peer, WG_PEER_UPDATE) call directly in order to upload the new psk for the peer to wireguard. Afterwards it should work on the next automatic handshake attempt.

jonasjelonek reviewed (commented):
addressed all comments, split up messages, removed unnecessary (for now) changes

nbd168 commented on psk-kex.c:
```diff
@@ -0,0 +1,368 @@
+	if (peer->kex_ctx.role != PSK_KEX_ROLE_INITIATOR)
+		return;
+
+	/* TBD: Can we send two messages in a row without issues? */
```
I would recommend splitting the sending of messages from initializing the key exchange state. Messages may have to be retransmitted over unreliable networks, and it would be good to not have to redo the keygen part each time.

I also think it might be useful to encrypt the initial messages with a key derived from curve25519 DH, which can be calculated once per peer ahead of time. That reduces the DoS potential of expensive PQ crypto, as long as the non-PQ crypto has not been cracked yet.
